### ISSUE TITLE
General feedback

### ISSUE THREAD

**Issue by indieSoftware** [scribd/Weaver#99]
Just wanted to provide some feedback / improvement thoughts after playing around with weaver 0.12 for one day. Btw, I like Weaver so far. It's a promising tool.

## Import statement

The readme mentions in the example for the `AppDelegate` code an import statement: `import Weaver`. However, when using Weaver via CocoaPods then there is no source code provided and thus nothing to import. Therefore trying to import something of course results into a `No such module 'Weaver'`.

## Generated files

When Weaver generates the boilerplate code then it's done in multiple files. That's fine, but I have to to add them every time. That's tedious when Weaver is part of the build phase, because that will result in the following kind of bad work flow:

1. Add a new weaver annotation to a file the first time.
2. Build the project (which runs weaver and creates a new weaver file for that file).
3. Building fails because some code couldn't be found (because the newly created files weren't added to the project, yet).
4. Adding the newly created file to the project and building again to proceed.

So even long after setting up weaver in the project I always have to remember to add these specific generated files each time I have added a new file to the dependency graph.

Therefore I think it would be better weaver creates all its code into one single file instead of multiple files. Maybe similar to what `R.swift` does. That way I could set up Weaver and add a `Weaver.swift` file to the project. Now every time weaver runs it just overwrites this single file and I never have to add something again related to Weaver. That will of course make that file really big, but in my opinion that's fine, because it's an auto-generated file anyway so it's not meant to be read and maintained by the developer. It simply has to work without any further management by a developer.

## initializer of container

When looking at the generated boilerplate code Weaver created I could find an `init()` method which initializes its properties.

Let's say I have two ViewControllers `VC1` and `VC2` with an annotation in VC1 like

```
// weaver: vc2 = VC2 <- UIViewController
// weaver: vc2.scope = .container
```

That will result in generated code like:

```
final class VC1DependencyContainer: VC1DependencyResolver {
	private var _vc2: UIViewController?
	var vc2: UIViewController {
		if let value = _vc2 { return value }
		let value = VC2()
		_vc2 = value
		return value
	}

	init() {
		_ = vc2
	}
}
```

Why is this `_ = vc2` in the `init` necessary? Isn't it also fine to have the controller being initialized lazily when it's really accessed the first time? Maybe it will never be needed but as soon as the dependency container is created this controller is already instantiated.

When VC2 also has dependencies to i.e. VC3 and so forth the whole view controller tree will be instantiated the first time the VC1 class is instantiated. Now let only one of those dependencies take some time when instantiating (i.e. when something resource have is created like a formatter or something like that), then there will be a performance issue at the beginning of VC1 even when VC3 with its heavy resource is never used.

Maybe I just got something wrong and oversaw something, but at the moment it just feels odd to have the objects instantiated in the init even when not yet used.

## Dependency matching

I had an interesting issue when trying to use a reference annotation. `VC1` looked like:

```
// weaver: vc2 = VC2 <- UIViewController
// weaver: vc2.scope = .container

// weaver: worker = Worker
// weaver: worker.scope = .container
```

A worker class `Worker` then tried to also depend on VC2 but not on it as generic UIViewController:

```
// weaver: vc2 <- VC2
```

That resulted an error in `extension VC1DependencyContainer: WorkerInputDependencyResolver {}`: `Type 'VC1DependencyContainer' does not conform to protocol 'WorkerInputDependencyResolver'`

That's true, because the protocol looks like:

```
protocol WorkerInputDependencyResolver {
	var vc2: VC2 { get }
}
```

While the container only provided the property for the generic type `UIViewController`:

```
final class VC1DependencyContainer: VC1DependencyResolver {
	var scene2VC: UIViewController {
	...
	}
```

So that was clearly an issue by me not matching the type of the registration and referencing annotation, however, Xcode of course didn't point me to my false annotation comment, but to a generated code file letting me guess why something didn't conform. It would be really nice if there is possibility to get my weaver annotations type checked so that Xcode points me to my comment instead (don't know if this is possible at all).

**Comment by trupin** [comment]
Hi @indieSoftware,

Thanks a lot for taking the time to write down your feedback, it's always helpful to see what developers like/dislike in Weaver.

> Import statement

Weaver doesn't include a framework anymore. It's a command line tool only. That's why there's nothing to import. I updated the README to reflect that since this `import Weaver` was clearly a left over from an old version of Weaver.

 > Generated files

Weaver actually already has the option [`--single-output`](https://github.com/scribd/Weaver#parameters) to make it generate everything in one file called `Weaver.swift`.

> initializer of container
> Why is this `_ = vc2` in the `init` necessary?

Lazily initializing dependency is problematic for two major reasons:
1. `lazy var`s aren't thread safe, which means dependency containers wouldn't be thread safe anymore if they were using `lazy var`s. I implemented thread safe lazy vars in one version of Weaver, but it's tricky and less efficient than building dependencies up front. Btw `lazy var`s are in general less efficient than regular regular `var`s and IMHO shouldn't be used unless it's an obvious fit to solve a specific problem.
2. Lazy instantiations dramatically change the initialization timing of dependencies. If dependencies where to be instantiated lazily, Weaver and developers would have a very hard time to predict when and why a dependency gets instantiated. With up front instantiations (which mimics how dependency injection would be done manually), the timing is clear. When instantiating a class, all its dependencies are instantiated upfront.

> Isn't it also fine to have the controller being initialized lazily when it's really accessed the first time? Maybe it will never be needed but as soon as the dependency container is created this controller is already instantiated.

IMO, if a dependency isn't going to be used, then it shouldn't be declared as a dependency. I imagine that there could be scenarios where a dependency could be used under certain conditions only, but I don't think this is enough to justify the use of `lazy var`s with all their downsides.

> Dependency matching
> It would be really nice if there is possibility to get my weaver annotations type checked so that Xcode points me to my comment instead (don't know if this is possible at all).

This is something which was asked before. I'll consider it for a future release.

**Comment by indieSoftware** [comment]
> Weaver actually already has the option `--single-output` to make it generate everything in one file called `Weaver.swift`.

Ah, that one I wasn't aware of, somehow I overlooked it.
So I tried it now, but somehow it doesn't work for me. When adding this flag no files were created at all.

> 1. `lazy var`s aren't thread safe

That's true and something I didn't think of. However, lazy vars solve one specific problem and that's deferring the initialization of cost intensive initializations. That's the reason why they exist. 

And the problem I currently see with the dependency graph is as mentioned that the whole tree is created at creation time of the container. Scenario: The app has 10 screens (view controllers and the controllers where to transition to are dependencies of their parent controller) then all 10 controllers plus their other possible dependencies are created before the first view appears even when the user maybe might never get pass that first screen. That's a waste of memory and start up time and definitively nothing anybody would do when injecting the controllers manually.

Anyway, it seems you also saw that problem because you already tried to implement thread safe lazy vars as you said. So maybe just keep it simple by introducing another annotation flag which will mark the dependency as being lazily created in the cost of not being thread safe? Reminds me of `nonatomic` in Objective-C. With that the developer has to care of any threading problems, but can defer some creation time when appropriate.

I would use that for example for injecting the view controllers needed for transitioning, which should only happen on a main thread so no threading problems, but I could defer the instantiation time with that flag.

**Comment by indieSoftware** [comment]
Hm, it seems the scope annotation `transient` is the appropriate one for view controller dependencies. You normally don't want to make a view controller a singleton, so `container` doesn't fit here. That also solves the problem with the dependency graph and instantiating all controllers up front.

Anyway, for other dependencies like workers or sort of I can imagine such a lazy var annotation would still make sense.

**Comment by trupin** [comment]
> multiple files were created, but not where they should be. So what am I doing wrong?

The `project_path` cannot be overridden in the configuration file since the configuration file is supposed to be located at the project path. Thus, Weaver takes the configuration path as the project path.

> Hm, it seems the scope annotation `transient` is the appropriate one for view controller dependencies.

Yes that's the scope I would use for your scenario.

> Anyway, for other dependencies like workers or sort of I can imagine such a lazy var annotation would still make sense.

I could indeed add a `lazy` scope and be explicit in the documentation that it's not thread-safe but my thinking is that it wouldn't be very useful since `DIContainers` take their dependencies built upfront, which means that a `lazy` dependency would get built as soon as another `DIContainers` requires it, which means in most cases, almost immediately.

> Doesn't the `weak` scope annotation do this? As far I can see it the `weak` annotation makes the dependency lazily loaded or am I wrong?

Yes, the `weak` scope implements a lazy loading because there's no way to implement a dependency upfront and store it in a `weak var`. It would get released immediately. This makes me think that I should document how using a `weak` scope makes the `DIContainer` not thread safe.

**Comment by indieSoftware** [comment]
> The `project_path` cannot be overridden in the configuration file since the configuration file is supposed to be located at the project path.

Ok, that works for me now, thanks. You really should point that out in the readme.

> I could indeed add a `lazy` scope

After playing more with weaver I now doubt a `lazy` scope would be necessary because the `weak` and `transient` scopes already do the trick. Especially with `weak` the container's holder just has to hold a strong reference to the weak dependency to solve what a lazy scope would do. 

> When I delete a previously created `Weaver.swift` file and then build the project the file gets created, but Xcode always cannot find that file on the first try. I always have to build twice to get passed even when my build script is above the `Compile Sources` phase.

Do you have a suggestion for this problem? Without that solved a CI system will always fail when the `weaver.swift` file is not part of the repo.

**Comment by indieSoftware** [comment]
I have some more issues I stumbled upon. Please have a look at the following code where I put all problems in one place:

```
class Root {
	let dependencies = RootDependencyContainer()

	// weaver: vc1 = VC1
	// weaver: vc1.scope = .transient

	// weaver: foo = Foo
	// weaver: foo.scope = .container

	// weaver: bar = Bar
	// weaver: bar.scope = .container

	func main() {
		_ = dependencies.vc1
	}
}

class VC1 {
	let dependencies: VC1DependencyResolver

	// weaver: foo <- Foo

	// weaver: bar <- BarInterface

	init(injecting dependencies: VC1DependencyResolver) {
		self.dependencies = dependencies

		let foo1 = dependencies.foo(foo: true)
		print("foo1.foo = \(foo1.foo)")
		let foo2 = dependencies.foo(foo: false)
		print("foo2.foo = \(foo2.foo)")
	}
}

protocol BarInterface {}

class Bar: BarInterface {}

class Foo {
	let dependencies: FooDependencyResolver

	// weaver: foo <= Bool

	init(injecting dependencies: FooDependencyResolver) {
		self.dependencies = dependencies
	}

	var foo: Bool {
		return dependencies.foo
	}
}
```

## 1. Abstraction for children

In the example code `Root` has a dependency to `bar` where bar gets created from `Bar` and is inside of `Root` also visible as this type. That's fine for `Root` because it creates it and thus knows the type anyway. However, VC1 doesn't need to know anything about `Bar`,  all it needs to know is `BarInterface`. Therefore I added the annotation `bar <- BarInterface` in VC1, but that doesn't work.

To solve this I have either to mark bar in `Root` as `bar = Bar <- BarInterface` or I have VC1 let to know `Bar` via `bar <- Bar`. The latter is obviously not the way to choose, but the former I also don't really like, because `Root` might depend really on `Bar`, not on `BarInterface`, e.g. `Root` might want to prepare or set up more on that instance of `Bar` which isn't part of the protocol.

Therefore it would be nice when it's possible to override the dependency for children by abstracting them down the tree.

## 2. Parameter Annotation

Using parameter annotations on dependencies with a `container` scope don't work.

In the example code the `foo` dependency in `Root` has a container scope thus it is created when the `RootDependencyContainer` gets created. However, the container's init method doesn't know anything about a `foo` parameter and therefore can't create an instance of `Foo`.

To solve this I first have to move the `foo` annotation from `Root` to `VC1` (replacing `foo <- Foo` there) and second I have to mark the scope to be `weak` or `transient`. Neither `container` nor `graph` work here and it has to be moved to `VC1` to get the compiler errors solved.

That's a big issue in my opinion, because that limits dependencies with parameters very much. Imagine `Foo` is kind of a server worker I want to initialize with a token the first time the user has logged in and the view controller with this dependency gets shown. The worker has to be accessible by all child containers, but that's currently not possible because it has to be registered in each child class or I have to set the parameter manually which would make weaver obsolete for this kind of dependency.

The issue with it continues when looking at the print statements. Set foo's scope to `weak` and put the annotation in VC1. When accessing the `foo` dependency I always have to provide a boolean parameter, even when it's not needed. The first call saved to `foo1` creates an instance and holds it with the strong reference, thus the provided `true` parameter is also correctly returned. However, when accessing `dependencis.foo` again I have to provide a parameter again, but that one is not used, thus `foo2.foo` also returns `true` even when provided `false` a line before. That's of course because the scope is `weak` and an instance already available. However, this most likely leads to errors later. Therefore I wouldn't recommend to make such parameterized property available as `weak`. A scope of `transient` on the other hand seems to be safe (as far as I can judge it).

**Comment by trupin** [comment]
> Do you have a suggestion for this problem? Without that solved a CI system will always fail when the `weaver.swift` file is not part of the repo.

This is probably something related to your project configuration. Weaver only generates the file and doesn't interact with Xcode. Not sure I can help more than that without more details.

> To solve this I have either to mark bar in `Root` as `bar = Bar <- BarInterface`.

That's how you're supposed to do it.

> The latter is obviously not the way to choose, but the former I also don't really like, because `Root` might depend really on `Bar`, not on `BarInterface`, e.g. `Root` might want to prepare or set up more on that instance of `Bar` which isn't part of the protocol.

That's an interesting feedback. We could easily make Weaver generate two accessors, one for the concrete type, and one with the abstract type. Something like this for example:

```swift
var _bar: Bar { ... }
var bar: BarProtocol { return _bar }
```

> Using parameter annotations on dependencies with a `container` scope don't work.

That's an expected behavior. Weaver can't know about which parameter you want to inject upfront, thus can't inject it automatically. This is something I should specify in the documentation. A dependency taking parameters at initialization can only be used as a transient dependency.

I get that this is very limiting and I have been thinking about ways to improve that. What I have in mind so far to improve that would be to add a way to pass parameters at the call site. For example:

```swift
// foo = Foo(foo: true)
```

Thank you for all of your feedback. I will think more about all these issues and file what I think we could solve.

**Comment by indieSoftware** [comment]
> This is probably something related to your project configuration.

I researched a little bit more about the build phase set up and found the following link enlightening:
https://indiestack.com/2014/12/speeding-up-custom-script-phases/

So to make sure weaver always creates the weaver.swift file I had to specify that file as `Output Files`. So in the build phase script I added something like `$(SRCROOT)/MyApp/Generated/Weaver.swift` to the `Output Files`. That will become tedious really fast when not using a single output file so I highly recommend using the `single_output` flag. After cleaning any derived data, closing and re-opening Xcode everything works as expected.

Now when deleting the `waver.swift` file and building the project Xcode doesn't throw any errors anymore and a CI build now should work fine. That's definitively something you should mention in the readme in the section how to set up the weaver build phase script.

> We could easily make Weaver generate two accessors

That sounds good. When the base class is always passed to all child containers it should be possible for any child class to only refer to a specific protocol that class conforms to. So when another class annotates `bar <- Boo` that should still work (provided `Bar` also conforms to the other protocol `Boo`).

I would really appreciate that change.

> A dependency taking parameters at initialization can only be used as a transient dependency.

Currently a `weak` dependency is also possible, but as described might lead to side effects and bugs. Therefore I would recommend removing the weak support for parameter annotations or at least note a warning about it in the documentation.

> Note that you can currently use a custom builder to pass static parameters

I think you forgot a `-> Foo` return type in your `make` method.

Anyway, this make approach has two very limiting restrictions.

1. That won't work anymore when `Foo` itself has some dependencies and weaver annotations.
2. Because of `make` being a static method there is no way to pass a variable as parameter, it has to be a constant or a static variable. The former is useless (at least I can't see a useful use case for that at the moment) and the latter error prone (thread safety, knowledge of instantiation time and setup).

By the way when `make` is put in `VC1` instead of `Foo` and the method looks like `static func make(_ res: VC1DependencyResolver) -> Foo { return res.foo }` it results in a recursive infinite loop during runtime without being noticed by weaver upfront. That's a simple example, but I'm sure we could create a more complex one which wouldn't be so obvious. Passing a resolver as parameter to an initializing method is dangerous in my opinion, especially because we are at that moment in the container's init method and thus not everything might have been initialized, yet. You already mentioned something like that in the readme by `Warning - Make sure you don't do anything unsafe with the DependencyResolver parameter passed down in this method since it won't be caught by the dependency graph validator.` but to be honest I didn't understand that warning when read the first time, because `unsafe` isn't very specific.

A use case how I would like to use a parameter dependency:

A `DetailViewController` depends on a `selectedIndex: Int` because it shows the details for a provided index selected by the user in the parent controller. However, the view controller only needs that parameter to pass it on a `BusinessLogic` which is another dependency of that view controller. The view controller is marked as `transient` because that's the way a view controller should be used as I have learned now. However, the `BusinessLogic` has a `container` scope so it gets created when the view controller gets created and also get referenced on many places in the view controller's code. So marking the parameter for the view controller doesn't seem problematic because it is `transient`, however for the `BusinessLogic` it is.

I don't want to mark the business logic to be transient because it holds some states and thus needs to be created only once for the lifetime of the view controller. It is also a real dependency of the view controller and used throughout the code.

However, after your hint with the make method I could mark `businessLogic.builder = ViewController.makeBL` and provide a make function like `static func makeBL(_ res: ViewControllerDependencyResolver) -> BusinessLogic { return BusinessLogic(selectedIndex: res.selectedIndex) }`. It seems that's the way how to use parameter dependencies for containers.

> `// foo = Foo(foo: fooValue)`

With my conclusion in the last section I now think I understand your approach here. I simply would write an annotation like `businessLogic = BusinessLogic(selectedIndex: selectedIndex)` in `ViewController` right after `selectedIndex <= Int`. That would make it obsolete to write that make function. I like it! 👍

**Comment by indieSoftware** [comment]
It seems the new approach without the make function is already implemented. Try out this:

```
class Root {
	let dependencies = RootDependencyContainer()

	// weaver: vc1 = VC1
	// weaver: vc1.scope = .transient

	func main() {
		_ = dependencies.vc1(val: true)
	}
}

class VC1 {
	let dependencies: VC1DependencyResolver

	// weaver: foo = Foo
	// weaver: foo.scope = .container

	// weaver: val <= Bool

	init(injecting dependencies: VC1DependencyResolver) {
		self.dependencies = dependencies

		let foo1 = dependencies.foo(val: false)
		print("foo1.val = \(foo1.val)")
	}
}

class Foo {
	let dependencies: FooDependencyResolver

	// weaver: val <= Bool

	init(injecting dependencies: FooDependencyResolver) {
		self.dependencies = dependencies
	}

	var foo: Bool {
		return dependencies.val
	}
}
```

Similar to the ViewController-BusinessLogic example here `VC1` and `Foo` classes have parameter annotations for `val` of type Bool. As long as the parameter's name is the same for both classes weaver could create valid code which compiles and it works even with a `container` scope for `foo`.

The only flaw is that each call of `dependencies.foo` needs a parameter passed for `val` which is never respected after the initial creation. `vc1` is created with the `val` flag set to `true` so no surprise that `dependencies.foo(val: false).val` in `VC1` still returns `true`.

Maybe a solution might be to simply make the signature of `foo` to be `var foo: Foo` for annotation scopes of `container` while keeping `func foo(val: Bool) -> Foo` for scopes of `transient`. All that remains is then to be sure the parameter is passed by the parent container and the parameter's name is the same, but the name equality is already a constraint for registration and reference annotations anyway.

**Comment by indieSoftware** [comment]
Btw, it would also be nice to have a configuration annotation to force weaver to inject some dependencies even when they are empty. For example:

```
class Foo {
	let dependencies: FooDependencyResolver

	// weaver: self.forceGenerate = true

	init(injecting dependencies: FooDependencyResolver) {
		self.dependencies = dependencies
	}
}
```

should result in

```
protocol FooDependencyResolver {
}

final class FooDependencyContainer: FooDependencyResolver {
	init() {
	}
}
```

The use case would be to prepare classes which have not yet set up properly with their dependencies. That might happen during development when preparing new dummy scenes to implement later while the init method should already match. Or when creating templates.

Without that annotation I have to either create a dummy dependency which has later to be removed or the init method has to be changed when introducing the first dependency. I would prefer everything is set up and ready to use weaver dependencies.

**Comment by indieSoftware** [comment]
Btw2, when something goes wrong while weaver is resolving the dependency graph some errors are emitted, but no nonzero exit code as Xcode mentions:
`Command PhaseScriptExecution emitted errors but did not return a nonzero exit code to indicate failure`

**Comment by indieSoftware** [comment]
When using the parameter annotation something strange happens when using the same parameter name again:

```
class Root {
	let dependencies = RootDependencyContainer()

	// weaver: vc = VC1
	// weaver: vc.scope = .transient

	func main() {
		_ = dependencies.vc(param: Param1())
	}
}

struct Param1 {}

class VC1 {
	let dependencies: VC1DependencyResolver

	// weaver: param <= Param1

	// weaver: vc = VC2
	// weaver: vc.scope = .transient

	init(injecting dependencies: VC1DependencyResolver) {
		self.dependencies = dependencies

		_ = dependencies.vc(param: Param2())
	}
}

struct Param2 {}

class VC2 {
	let dependencies: VC2DependencyResolver

	// weaver: param <= Param2

	init(injecting dependencies: VC2DependencyResolver) {
		self.dependencies = dependencies
	}
}
```

Here the code won't compile because of `_ = dependencies.vc(param: Param2())` in `VC1`. `param` of `VC2` is declared as of type `Param2`, and providing this parameter doesn't work because weaver keeps that parameter to be `Param1`.

To solve this I have to rename `param` in `VC2` to something else, i.e. `param2`. Then everything works. That limits the usage of names in the graph. That's especially bad when using Xcode templates. There my parameter variable is always named the same, but has different types according to the scene, so that won't work for me anymore.

Maybe a solution might be to restrict the passing of parameters by weaver only to container scopes, but not to transient scopes? So passing `param` to `Foo` which has a scope of `container` would be ok, because it's otherwise not possible to provide a parameter to Foo. However, `VC2` has a scope of `transient` and thus passing the parameter via the resolver is not necessary and thus shouldn't be related to any parameter declared prior in the graph.

**Comment by indieSoftware** [comment]
I have another issue. What's wrong with this code?

```
class Root {
	let dependencies = RootDependencyContainer()

	// weaver: vc1 = VC1
	// weaver: vc1.scope = .transient

	func main() {
		_ = dependencies.vc1
	}
}

class VC1 {
	let dependencies: VC1DependencyResolver

	// weaver: nav = Nav1
	// weaver: nav.scope = .container

	// weaver: log = Log1
	// weaver: log.scope = .container

	init(injecting dependencies: VC1DependencyResolver) {
		self.dependencies = dependencies
	}
}

class Nav1 {
	let dependencies: Nav1DependencyResolver

	// weaver: vc2 = VC2
	// weaver: vc2.scope = .transient

	init(injecting dependencies: Nav1DependencyResolver) {
		self.dependencies = dependencies
	}
}

class Log1 {
	let dependencies: Log1DependencyResolver

	// weaver: nav <- Nav1

	init(injecting dependencies: Log1DependencyResolver) {
		self.dependencies = dependencies
	}
}

class VC2 {
	let dependencies: VC2DependencyResolver

	// weaver: log = Log2
	// weaver: log.scope = .container

	init(injecting dependencies: VC2DependencyResolver) {
		self.dependencies = dependencies
	}
}

class Log2 {
	let dependencies: Log2DependencyResolver

	// weaver: forceGenerate = Bool

	init(injecting dependencies: Log2DependencyResolver) {
		self.dependencies = dependencies
	}
}
```

Weaver complains: `Detected invalid dependency graph starting with `nav: Nav1` in `Log1`.

I can solve this by renaming `log` in `VC2` into `log2` so it seems it's a naming problem again. However, the error message is totally misleading in this situation. And of course I don't want to rename my property.

**Comment by indieSoftware** [comment]
Just another thought:

When using weaver I have to write all the time something like `dependecies.XXX`. Wouldn't it be nice to have weaver also create some extensions to make accessing the dependencies a little bit easier?

For example we have a class with the following annotations:

```
class Root {
	let dependencies = RootDependencyContainer()

	// weaver: vc1 = VC1
	// weaver: vc1.scope = .transient
	// weaver: vc1.accessor = dependencies

	func main() {
		_ = vc1
	}
}
```

And weaver would also create an extension for `Root`:

```
extension Root {
	private var vc1: VC1 { return dependencies.vc1 }
}
```

That way we can simply treat the dependency `vc1` as it were a property and can write something like `_ = vc1` instead of `_ = dependencies.vc1`.

If it's also possible to determine the container's or resolver's property name automatically (here `dependencies: RootRependencyContainer`) then we even wouldn't need the annotation `// weaver: vc1.accessor = dependencies` anymore. The extension then could be generated automatically per default, only an annotation like `// weaver: vc1.accessor = off` would prevent this or maybe a global flag.

**Comment by trupin** [comment]
Hi @indieSoftware,

Sorry for the late answer, I've been pretty busy lately and there's a lot to cover in your last messages. 

> So to make sure weaver always creates the weaver.swift file I had to specify that file as Output Files. So in the build phase script I added something like $(SRCROOT)/MyApp/Generated/Weaver.swift to the Output Files. That will become tedious really fast when not using a single output file so I highly recommend using the single_output flag. After cleaning any derived data, closing and re-opening Xcode everything works as expected.

There's a [specific command to](https://github.com/scribd/Weaver#generate-xcfilelist-files) generate the Output/Input `xcfilelist` files.

> Now when deleting the weaver.swift file and building the project Xcode doesn't throw any errors anymore and a CI build now should work fine. That's definitively something you should mention in the readme in the section how to set up the weaver build phase script.

I fixed this on master. I'll be pushing a new release with that fix in asap.

> You already mentioned something like that in the readme by Warning - Make sure you don't do anything unsafe with the DependencyResolver parameter passed down in this method since it won't be caught by the dependency graph validator. but to be honest I didn't understand that warning when read the first time, because unsafe isn't very specific.

I will try to make that part more clear in the documentation.

> The only flaw is that each call of dependencies.foo needs a parameter passed for val which is never respected after the initial creation. vc1 is created with the val flag set to true so no surprise that dependencies.foo(val: false).foo in VC1 still returns true.

I completely agree, that's the same issue which has been mentioned in #102. As I answered there, I will work on improving that part of Weaver. 

> Weaver complains: Detected invalid dependency graph starting with 'nav: Nav1' in 'Log1'.
> I can solve this by renaming log in VC2 into log2 so it seems it's a naming problem again. However, the error message is totally misleading in this situation. And of course I don't want to rename my property.

Yes, this is another known limitation with Weaver. Since it resolves dependencies based on their name, and since names are considered global to the project, It is not supported to use the same name with two different types in the same dependency graph. A profound change in the way Weaver resolves dependencies is needed to support this. I have been wanting to make that change for a while but I couldn't find the right approach so far. I think I finally came up with something reasonable and will start implementing it soon.

> When using weaver I have to write all the time something like dependecies.XXX. Wouldn't it be nice to have weaver also create some extensions to make accessing the dependencies a little bit easier?

I agree with you that it would be nice, but it would break encapsulation of dependencies by making all of them public. IMO, it's usually when developers start doing this that bad things start happening in a project. To keep the dependencies private we would need to be able to privately implement a protocol containing the accessors' helpers, which isn't supported by Swift at the moment. Thus, I think we'll have to stick with using the `private dependencies` property for now.

I realize that Weaver has few serious limitations at the moment, and that it's not fair that its API looks like it is supporting parameters (for example), but that it's not actually fully supported under the hood. I will make sure to warn developers about these limitations in the documentation until I improve Weaver. Also feel free to open a PR to fix the documentation wherever you think there's a lack of information, I'll be pleased to review it and merge it into master.

Also, I am thinking of rewriting Weaver to make all these improvements at once. Some of them need to change the way Weaver works at a low level and I doubt simple fixes will be enough. Your feedback has been really helpful to point out what I need to focus on for this rewrite so thank you for that. I plan to be a little more committed to this project in the next few months in order to make this rewrite happen and I hope it'll fix most of the limitations you've mentioned here.

**Comment by indieSoftware** [comment]
> Sorry for the late answer, I've been pretty busy lately

No problem, mostly we are all pretty busy.

> There's a specific command to generate the Output/Input xcfilelist files.

Seems like a good alternative.

> Thus, I think we'll have to stick with using the private dependencies property for now.

I totally agree with you.

> Your feedback has been really helpful to point out what I need to focus on for this rewrite so thank you for that. I plan to be a little more committed to this project in the next few months in order to make this rewrite happen and I hope it'll fix most of the limitations you've mentioned here.

You're welcome. I'm really interested in a new version and can't wait to try it in my current project. So feel free to contact me when you have released a new version (or are about to) and would like any further feedback.
